### PR TITLE
prov/net: Initial integration of io_uring to the provider

### PR DESCRIPTION
**Pull request by sydidelot**

### CODE REVIEW

File: prov/net/src/xnet_progress.c [hunk 1]
```diff
@@ -783,6 +783,37 @@ static void xnet_run_ep(struct xnet_ep *ep, bool pin, bool pout, bool perr)
...
 
+static void xnet_progress_cqe(struct xnet_uring *uring,
+			      ofi_io_uring_cqe_t *cqe)
+{
+	struct ofi_bsock *bsock;
+
+	assert(xnet_io_uring);
+	bsock = (struct ofi_bsock *) cqe->user_data;
+	assert(bsock);
+}
```
Comment (shefty) [inline prov/net/src/xnet_progress.c:810]: I'm hoping a future patch expands this function to do more, yes?

**Review by shefty — COMMENTED** [review]
Looks good.  Minor comments only.

File: prov/net/src/xnet_progress.c (hunk 1 above)
Comment (sydidelot) [reply prov/net/src/xnet_progress.c:810]: Yes, this is correct. I will be adding the code that completes and progresses TX/RX operations in a future PR.

File: prov/net/src/xnet_progress.c [hunk 2]
```diff
@@ -1101,6 +1157,14 @@ void xnet_close_progress(struct xnet_progress *progress)
...
 	assert(slist_empty(&progress->event_list));
 	xnet_stop_progress(progress);
+	if (xnet_io_uring) {
+		ofi_dynpoll_del(&progress->epoll_fd,
+				ofi_uring_get_fd(&progress->rx_uring.ring));
+		xnet_destroy_uring(&progress->rx_uring);
+		ofi_dynpoll_del(&progress->epoll_fd,
+				ofi_uring_get_fd(&progress->tx_uring.ring));
+		xnet_destroy_uring(&progress->tx_uring);
+	}
```
[the remark this answers is not in the document]
Comment (sydidelot) [reply prov/net/src/xnet_progress.c:1226]: Done. Thanks

File: prov/net/src/xnet_progress.c [hunk 3]
```diff
@@ -1081,8 +1103,42 @@ int xnet_init_progress(struct xnet_progress *progress, struct fi_info *info)
...
+				      xnet_default_rx_size);
+		if (ret)
+			goto err7;
+
+		ret = ofi_dynpoll_add(&progress->epoll_fd,
+				      ofi_uring_get_fd(&progress->rx_uring.ring),
+				      POLLIN, &progress->rx_uring.fid);
+		if (ret)
+			goto err8;
+	}
```
[the remark this answers is not in the document]
Comment (sydidelot) [reply prov/net/src/xnet_progress.c:1189]: Done.

**Comment by shefty** [comment]
AWS CI failure is related to 1 EFA test.

**Comment by shefty** [comment]
thanks!

**Pull request merged**